04-cs-fundamentals/internship/hash-tables.md

```diff
@@ -0,0 +1,319 @@
+
+Now that we've had some theory, when would you expect to use a hash table?  In any problem where quick lookup of items is important, a hash table should be considered.  However a hash table depends on **unique keys**.  If a dataset cannot provide unique key values, a hash table may not be the best solution.  Instead a binary search tree or another data structure may be a better selection.
+
+### Sample Problem
+
+Hash tables are a common, practical, solution to a variety of programming problems.  Below is a sample interview problem taking from [leetcode.com](https://leetcode.com/problems/sort-characters-by-frequency/)
+
+Take a look at the problem below and answer:
+
+<details>
+  <summary>Why would a hash table be a good approach here?</summary>
+
+  - To solve this problem we need to store each letter and the number of occurances. <br />
+    - There are a few of ways to do this:
+      1. We could traverse the list repeatably counting the ocurances of each letter.  This would require nested loops and O(n<sup>2</sup>) time complexity.
+      1. We could also sort the string by letter use this sorted string to count the occurances of each letter.  Sorting the string would require O(nlogn) time complexity.
+      1. Or we could use a hash table to with each character as the keys and the number of occurances as the value.  Building this hash would require one traversal of the string or O(n).  
+    - Once we have the list of letters and their number of occurances, we can use this to build the output string.
+
+  Because looking up an items in a hash table is so fast, hash tables turn out to be excellent data structures.
+
+</details>
+<br />
+
+#### Exercise
+
+With a neighbor examine the problem below and do the following:
+
+1.  Write down 2-3 clarifying questions.
+1.  Write down assumptions you will make for each clarifying question.
+1.  Generate a fresh set of 2-3 sample input-output looking for **edge cases**.
+1.  Break down the problem into at least 2 subproblems, more if you can.
+1.  Write out a pseudocode solution.
+
+**Problem Description**
+
+Given a string, sort it in decreasing order based on the frequency of characters.
+
+**Example 1:**
+
+```
+Input:
+"tree"
+
+Output:
+"eert"
+
+Explanation:
+'e' appears twice while 'r' and 't' both appear once.
+So 'e' must appear before both 'r' and 't'. Therefore "eetr" is also a valid answer.
+Example 2:
+
+Input:
+"cccaaa"
+
+Output:
+"cccaaa"
+```
+
+**Explanation:**
+Both 'c' and 'a' appear three times, so "aaaccc" is also a valid answer.
+Note that "cacaca" is incorrect, as the same characters must be together.
+
+**Example 3:**
+
+```
+Input:
+"Aabb"
+
+Output:
+"bbAa"
+```
+
+**Explanation:**
+
+"bbaA" is also a valid answer, but "Aabb" is incorrect.
+Note that 'A' and 'a' are treated as two different characters.
+
+You can find a [solution in the example code folder](example-code/character_frequency.rb).
+
+## Array, Linked List, Binary Search Tree or Hash Table
```

Comment:
I'd kind of like some questions where array or linked list are the clear best answer.  Though I don't have any good real life situations off the top of my head.